src/Data/Parameterized/FinMap/Safe.hs

```diff
@@ -0,0 +1,209 @@
+Copyright        : (c) Galois, Inc 2022
+
+See "Data.Parameterized.FinMap".
+-}
+
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Data.Parameterized.FinMap.Safe
+  ( FinMap
+  -- * Query
+  , null
+  , lookup
+  , size
+  -- * Construction
+  , incMax
+  , embed
+  , empty
+  , singleton
+  , insert
+  , buildFinMap
+  , append
+  , fromVector
+  -- * Operations
+  , delete
+  , decMax
+  , mapWithKey
+  ) where
+
+import           Prelude hiding (lookup, null)
+
+import           Data.Foldable.WithIndex (FoldableWithIndex(ifoldMap))
+import           Data.Functor.WithIndex (FunctorWithIndex(imap))
+import           Data.Maybe (isJust)
+import           Data.Proxy (Proxy(Proxy))
+import           Data.Map (Map)
+import qualified Data.Map as Map
+import           GHC.TypeLits (KnownNat, Nat)
+
+import           Data.Parameterized.Fin (Fin)
+import qualified Data.Parameterized.Fin as Fin
+import           Data.Parameterized.NatRepr (NatRepr, type (+), type (<=))
+import qualified Data.Parameterized.NatRepr as NatRepr
+import           Data.Parameterized.Vector (Vector)
+import qualified Data.Parameterized.Vector as Vec
+
+------------------------------------------------------------------------
+-- Type
+
+-- | @'FinMap' n a@ is a map with @'Fin' n@ keys and @a@ values.
+data FinMap (n :: Nat) a =
+  FinMap
+    { getFinMap :: Map (Fin n) a
+    , maxSize :: NatRepr n
+    }
+
+instance Eq a => Eq (FinMap n a) where
+  fm1 == fm2 = getFinMap fm1 == getFinMap fm2
+  {-# INLINABLE (==) #-}
+
+instance Functor (FinMap n) where
+  fmap f fm = fm { getFinMap = fmap f (getFinMap fm) }
+  {-# INLINABLE fmap #-}
+
+instance Foldable (FinMap n) where
+  foldMap f = foldMap f . getFinMap
+  {-# INLINABLE foldMap #-}
+
+instance FunctorWithIndex (Fin n) (FinMap n) where
+  imap f fm = fm { getFinMap = Map.mapWithKey f (getFinMap fm) }
+  -- Inline so that RULES for Map.mapWithKey can fire
+  {-# INLINE imap #-}
+
+instance FoldableWithIndex (Fin n) (FinMap n) where
+  ifoldMap f = Map.foldMapWithKey f . getFinMap
+  {-# INLINABLE ifoldMap #-}
+
+-- | Non-lawful instance, provided for testing
+instance Show a => Show (FinMap n a) where
+  show fm = show (getFinMap fm)
+  {-# INLINABLE show #-}
+
+------------------------------------------------------------------------
+-- Query
+
+-- | /O(1)/. Is the map empty?
+null :: FinMap n a -> Bool
+null = Map.null . getFinMap
+{-# INLINABLE null #-}
+
+-- | /O(log n)/. Fetch the value at the given key in the map.
+lookup :: Fin n -> FinMap n a -> Maybe a
+lookup k = Map.lookup k . getFinMap
+{-# INLINABLE lookup #-}
+
+-- | /O(nlog(n))/. Number of elements in the map.
+size :: forall n a. FinMap n a -> Fin (n + 1)
+size fm =
+  Fin.countFin (maxSize fm) (\k _count -> isJust (lookup (Fin.mkFin k) fm))
```

Comment:
Could this be `O(n)` if implemented as  `length $ filter isJust (snd <$> Map.elems (getFinMap fm))`, assuming a `Fin n` could be directly constructed from the `Int` obtained?